pkg/keycloak/realm/phaseHandler.go

```diff
@@ -0,0 +1,449 @@
+		return kcr, err
+	}
+
+	realm, err := kcClient.GetRealm(kcr.Name)
+	if err != nil {
+		return kcr, errors.Wrap(err, "error retrieving realms from keycloak")
+	}
+
+	if realm != nil && kcr != nil && realm.Spec.Realm == kcr.Spec.Realm {
+		kcr.Status.Phase = v1alpha1.PhaseProvisioned
+		return kcr, nil
+	}
+
+	storeUsers := kcr.Spec.KeycloakApiRealm.Users
+	storeClients := kcr.Spec.KeycloakApiRealm.Clients
+
+	kcr.Spec.KeycloakApiRealm.Clients = []v1alpha1.KeycloakClient{}
+	kcr.Spec.KeycloakApiRealm.Users = []v1alpha1.KeycloakUser{}
+
+	err = kcClient.CreateRealm(kcr)
+
+	kcr.Spec.KeycloakApiRealm.Clients = storeClients
+	kcr.Spec.KeycloakApiRealm.Users = storeUsers
+
+	if err != nil {
+		return kcr, errors.Wrap(err, "error creating keycloak realm")
+	}
+
+	kcr.Status.Phase = v1alpha1.PhaseProvisioned
+
+	return kcr, nil
+}
+
+func (ph *phaseHandler) Provisioned(kcr *v1alpha1.KeycloakRealm) (*v1alpha1.KeycloakRealm, error) {
+	kcClient, err := ph.getClient(kcr)
+	if err != nil {
+		return kcr, errors.Wrapf(err, "error reconciling keycloak realm: '%v'", kcr.Spec.Realm)
+	}
+
+	if err := ph.reconcileUsers(kcClient, *kcr, kcr.ObjectMeta.Namespace); err != nil {
+		return nil, err
+	}
+
+	if err := ph.reconcileClients(kcClient, *kcr, kcr.ObjectMeta.Namespace); err != nil {
+		return nil, err
+	}
+
+	if err := ph.reconcileIdentityProviders(kcClient, *kcr); err != nil {
+		return nil, err
+	}
+
+	return kcr, nil
+}
+
+func (ph *phaseHandler) reconcileUsers(kcClient keycloak.KeycloakInterface, realm v1alpha1.KeycloakRealm, ns string) error {
+
+	users, err := kcClient.ListUsers(realm.Spec.Realm)
+	if err != nil {
+		return err
+	}
+
+	kcUsers := map[string]*v1alpha1.KeycloakUser{}
+	for i := range users {
+		kcUsers[users[i].UserName] = users[i]
+	}
+
+	userPairsList := map[string]*v1alpha1.KeycloakUserPair{}
+	for i := range realm.Spec.Users {
+		user := &realm.Spec.Users[i]
+		userPairsList[user.UserName] = &v1alpha1.KeycloakUserPair{
+			SpecUser: user,
+			KcUser:   kcUsers[user.UserName],
+		}
+	}
+
+	for i := range userPairsList {
+		err := ph.reconcileUser(userPairsList[i].KcUser, userPairsList[i].SpecUser, realm.Spec.Realm, kcClient, ns)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (ph *phaseHandler) reconcileUser(kcUser, specUser *v1alpha1.KeycloakUser, realmName string, authenticatedClient keycloak.KeycloakInterface, ns string) error {
+	// create the user
+	// generate a password and update the users pass
+	// output the result to a secret
+	if kcUser == nil {
+		err := authenticatedClient.CreateUser(specUser, realmName)
+		if err != nil {
+			return err
+		}
+		// generate and update password
+		u, err := authenticatedClient.FindUserByEmail(specUser.UserName, realmName)
+		if err != nil {
+			return errors.Wrap(err, "failed find user")
+		}
+		if u == nil {
+			return errors.New("failed to find user " + specUser.Email)
+		}
+		var newPass string
+		if specUser.Password != nil {
+			newPass = *specUser.Password
```

Comment:
it's not required in the CRD, but I've added it as an option so the CR can specify a password, instead of waiting for a random password, this was added primarily to deal with the installer wanting the passwords to all be "Password1".